xn2v/csf_graph/csf_graph_tf.py

```diff
@@ -0,0 +1,395 @@
+            dest_idx: The index of a destination node.
+
+        Returns:
+            The weight of the edge that exists between the source and destination nodes.
+        """
+
+        for i in range(self.offset_to_edge_[source_idx], self.offset_to_edge_[source_idx + 1]):
+            if dest_idx == self.edge_to[i]:
+                return self.edge_weight[i]
+
+    def neighbors(self, source: str):
+        """Gets a list of node names, which are the neighbors of the user-provided source node.
+
+        Args:
+            source: The name of a source node.
+
+        Returns:
+            neighbors: A list of neighbors (i.e. strings representing node names) for a source node.
+        """
+
+        neighbors = []
+        source_idx = self.node_to_index_map[source]
+
+        for i in range(self.offset_to_edge_[source_idx], self.offset_to_edge_[source_idx + 1]):
+            nbr = self.index_to_node_map[self.edge_to[i]]
+            neighbors.append(nbr)
+
+        return neighbors
+
+    def neighbors_as_ints(self, source_idx: int):
+        """Gets a list of node indices, which are the neighbors of the user-provided source node.
+
+        Args:
+            source_idx: The index of a source node.
+
+        Returns:
+            neighbors_ints: A list of indices of neighbors for a source node.
+        """
+
+        neighbors_ints = []
+
+        for i in range(self.offset_to_edge_[source_idx], self.offset_to_edge_[source_idx + 1]):
+            nbr = self.edge_to[i]
+            neighbors_ints.append(nbr)
+
+        return neighbors_ints
+
+    def has_edge(self, src: str, dest: str):
+        """Checks if an edge exists between src and dest node.
+
+        Args:
+            src: The name of a source node.
+            dest: The name of a destination node.
+
+        Returns:
+            A boolean value is returned, where True means an exists and False means no edge exists.
+        """
+
+        source_idx = self.node_to_index_map[src]
+        dest_idx = self.node_to_index_map[dest]
+
+        for i in range(self.offset_to_edge_[source_idx], self.offset_to_edge_[source_idx + 1]):
+            nbr_idx = self.edge_to[i]
+
+            if nbr_idx == dest_idx:
+                return True
+
+        return False
+
+    @staticmethod
+    def same_nodetype(n1: str, n2: str):
+        """Encodes the node type using the first character of the node label. For instance, g1 and g2 have the same
+        node type but g2 and p5 do not.
+
+        Args:
+            n1: A string containing an node name.
+            n2: A string containing an node name.
+
+        Returns:
+            A boolean value is returned, where True means the user-provided nodes are of the same type and False
+            means that they are of different types.
+        """
+
+        if n1[0] == n2[0]:
+            return True
+        else:
+            return False
+
+    def edges(self) -> Union[tf.data.Dataset, tf.RaggedTensor]:
```

Comment:
@callahantiff do we need to allow `tf.RaggedTensor`s? IIUC, all edges will be of length 2 in `edges()` and `edges_as_ints()` right?

Reply:
While the code we current have does not need to account for `tf.RaggedTensors`, I talked to Peter about it and for future iterations and/or as new path finding or embedding methods are added we don't want to be constrained to lists or paths of the same lengths.

This is why I am proceeding with the refactor with the assumption that we should be able to accommodate either list of paths of equal length (`tf.data.Dataset`) of lists of paths of unequal lengths (`tf.RaggedTensor`). Does that make sense?

Are you running into a problem?

Reply:
Ah okay, thanks! That makes sense, no problem supporting `RaggedTensors`, just trying to understand the code better

Reply:
Sounds good. Is our plan still the same as Friday? I will focus on the `word2vec_tf.py` script and tests and you are working on the `xn2v/hetnode2vec_tf.py` script and tests?

I also noticed there are a bunch of merge conflicts, I'm happy work on those later today.

Reply:
Yep, I'm working on `xn2v/hetnode2vec_tf.py` - hope to have some more code soon

Reply:
Sounds Great! 😎